### PR TITLE
U5 support

### PR DESCRIPTION
**Pull request by tgree**
This pull request adds support for the STM32U5 microcontrollers.  These low-power MCUs are different from the other STM32 MCUs in that they do not have DMA or DMAMUX peripherals; instead they have GPDMA and LPDMA peripherals that are significantly different from their predecessors.  modm-devices previously assumed that all MCUs will have DMA peripherals so part of this changeset makes that optional.  If you look at the stm_device_tree.py diff while *ignoring whitespace changes* then you will see that the 224-line change is actually trivial.

Making DMA peripherals optional is possibly the right thing to do since one could argue that the STM32U5 doesn't have DMA/DMAMUX peripherals and therefore anything you were generating to support those peripherals is incompatible with the STM32U5.

I added U5 support because my company will be migrating to the U5 MCU soon and wanted to get the stm_layout tool working with the U5 so that we could plan our pin allocations and then noticed that modm-devices hadn't been updated to support U5 yet.  If you think this should be done in a completely different way then let me know how I can help!

### CODE REVIEW

**Comment by tgree** [comment]
I did do a full "make generate-stm32" after deleting everything from devices/stm32/* and the resulting diff was just the single new XML file with the U5 bits in it, so everything else is still generated exactly as before.

**Comment by tgree** [comment]
Looks like I missed updating the docs with the new device counts; done in the latest commit.  Looks like maybe the checkin tests also need to be updated somewhere to test the U5 files but I'm not sure where that lives.

File: devices/stm32/stm32u5-75_85.xml [hunk 1]
```diff
@@ -0,0 +1,3064 @@
...
+      <instance value="1"/>
+      <instance value="2"/>
+      <instance value="3"/>
+      <instance value="4"/>
+    </driver>
+    <driver name="icache" type="stm32-v1.0"/>
+    <driver name="irtim" type="stm32-v1.0"/>
+    <driver name="iwdg" type="stm32"/>
+    <driver name="levelx" type="stm32-v1.0"/>
+    <driver name="linkedlist" type="stm32-v1.0"/>
```
Comment (salkinium) [inline devices/stm32/stm32u5-75_85.xml]: It seems to me like some of these peripherals are actually CubeMX middleware names (those ending with x). Could you check and ignore them?

**Comment by salkinium** [comment]
The U5 job needs to be added here: https://github.com/modm-io/modm-devices/blob/develop/.github/workflows/tests.yml#L76

**Comment by salkinium** [comment]
I agree on the DMA data, we'll deal with it once we need it.

**Comment by salkinium** [comment]
Also bump the version to 0.4.0 here, so that the PyPi package gets updated: https://github.com/modm-io/modm-devices/blob/develop/modm_devices/__init__.py#L19

**Comment by tgree** [comment]
I see the following middlewares in the XML: FileX, LevelX, ThreadX, USBX.

I randomly Googled the LevelX one and it is a library for wear-levelling on flash devices, so I think you're right that these are just STM32 libraries.

I ignored them by adding them to the "software_ips" list; I think this is the right way but if there is another way that's better let me know and I'll fix it.  I've also added the U5 workflows test and bumped the version as requested.

After these changes I again did a full "make generate-stm32" and the only diff was the 4 lines of removed middleware from the generated U5 XML file.

**Comment by tgree** [comment]
Looks like I forgot to post my changes to the cmsis-header-stm32 repo; pull request submitted.  It also looks like the tests are now waiting for job stm32-f0-f1-f2-f3-f7-l1-wb-wl to complete; I renamed this to stm32-f0-f1-f2-f3-f7-l1-wb-wl-u5 in the yaml file so I'm not sure where its picking up the old name from.  git grep doesn't show anything in my modm-devices repo so maybe its something configured on the github side of things?

**Comment by salkinium** [comment]
> something configured on the github side of things?

Yes, I've changed it.

> I ignored them by adding them to the "software_ips" list;

But wait, there's more: linkedlist, netxduo and debug (not software, but still redundant).

**Comment by tgree** [comment]
Hmmm. Is there a way to trigger re-running the pull_request checks without updating a commit?  I suppose I could touch a timestamp and push again.

**Comment by tgree** [comment]
Oh, sorry I missed your other commend about linkedlist, netxduo and debug.  Addressing now.

**Comment by tgree** [comment]
How did you even spot those!?  Anyways, it looks like the DEBUG component shows up in a number of families already: H7, L5 (not present in modm-devices yet either it looks like), MP1 (probably out of scope), U5 and WL.  It isn't really clear to me what that DEBUG IP is since it doesn't have a corresponding ConfigFile entry with more info.

In the modm-devices generated XML files, they are showing up with what appears to be a generic type of "stm32-v1.0", however in the ST XML files they have different Version tags (i.e. "STM32WL_sys_v1_0", "STM32L55_sys_v1_0", etc).  I'm not sure if modm-devices is replacing "<FAMILY>_sys_" with "stm32-" and converting "v1_0" to "v1.0" or if something else is going on.

I'll make the change to blacklist DEBUG devices, I just wonder if there is actually some bit of hardware there?  It can't be the DBGMCU since there aren't DEBUG sections on all of the existing devices and I think DBGMCU is pretty much everywhere.  I know the Cortex-M33 found on U5 does have new debug features like micro-trace but I don't think that would apply to, say, the WL chips.  Do you have any ideas about what DEBUG may be?

**Comment by tgree** [comment]
Oh, I lied.  There are corresponding config files for DEBUG in the IP directory, they just aren't linked to the XML via a ConfigFile="..." tag.

**Comment by tgree** [comment]
At least the checks pass now!  Make sure that the last commit about updating device files is what you want since it does touch a bunch of generated XML that is unrelated to adding U5 support.

**Comment by salkinium** [comment]
>  Make sure that the last commit about updating device files is what you want since it does touch a bunch of generated XML that is unrelated to adding U5 support.

Yeah, I don't want that. I didn't realize there were GPIO signals attached to the Debug module (kinda obvious in hindsight), I thought it was only the internal Cortex-M peripherals.

I've squashed your changes and removed the DEBUG removal.

**Pull request merged**